p2p/libp2p/discovery/kadDhtDiscoverer2.go

```diff
@@ -0,0 +1,245 @@
+package discovery
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/ElrondNetwork/elrond-go/p2p/libp2p"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	opts "github.com/libp2p/go-libp2p-kad-dht/opts"
+	kbucket "github.com/libp2p/go-libp2p-kbucket"
+)
+
+// KadDhtDiscoverer is the kad-dht discovery type implementation
+// This implementation does not support pausing and resuming of the discovery process
+type KadDhtDiscoverer2 struct {
+	mutKadDht     sync.RWMutex
+	kadDHT        *dht.IpfsDHT
+	refreshCancel context.CancelFunc
+
+	contextProvider *libp2p.Libp2pContext
+
+	refreshInterval  time.Duration
+	randezVous       string
+	initialPeersList []string
+}
+
+// NewKadDhtPeerDiscoverer2 creates a new kad-dht discovery type implementation
+// initialPeersList can be nil or empty, no initial connection will be attempted, a warning message will appear
+func NewKadDhtPeerDiscoverer2(
+	refreshInterval time.Duration,
+	randezVous string,
+	initialPeersList []string) *KadDhtDiscoverer2 {
+
+	isListNilOrEmpty := len(initialPeersList) == 0
```

Comment:
this discoverer does little when initialPeerList is 0. Do we really need this discoverer ? couldn't we use the one from libP2P ?

Reply:
will this be able to discover anything without the initialPeerList?

Reply:
We need this option for an empty list for the first seeder (advertiser, in our integration tests) to "create" the network. Others will connect to this seeder and through the seeder's kad dht that is running in the background are able to see each-other. So we definitely need this mechanism.